{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "    source('./comparison/third_parties/symphony/vignettes/libs.R') # imports\n",
    "    source('./comparison/third_parties/symphony/vignettes/utils.R') # color definitions and plotting functions\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:future':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from 'package:plyr':\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Matrix':\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:plyr':\n",
      "\n",
      "    desc\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from 'package:data.table':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Biobase'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'AnnotationDbi'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"multiple methods tables found for 'aperm'\"\n",
      "Warning message:\n",
      "\"replacing previous import 'BiocGenerics::aperm' by 'DelayedArray::aperm' when loading 'SummarizedExperiment'\"\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'AnnotationFilter'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:future':\n",
      "\n",
      "    value\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'XVector'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:plyr':\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "Warning message:\n",
      "\"no function found corresponding to methods exports from 'BSgenome' for: 'releaseName'\"\n"
     ]
    }
   ],
   "source": [
    "source(\"./preprocess/utils.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(symphony)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cat_data <- function(rna, adt){\n",
    "    counts <- rbind(rna[['rna']]@counts, adt[['adt']]@counts)\n",
    "    adata <- CreateSeuratObject(counts)\n",
    "    return(adata)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"p1_0\" \"p2_0\" \"p3_0\" \"p4_0\" \"p5_0\" \"p6_0\" \"p7_0\" \"p8_0\"\n"
     ]
    }
   ],
   "source": [
    "config <- parseTOML(\"./configs/data.toml\")[['wnn_cellmask']]\n",
    "subset_names <- basename(config$raw_data_dirs[1:8])\n",
    "print(subset_names)\n",
    "subset_ids <- sapply(seq_along(subset_names) - 1, toString)\n",
    "input_dirs <- pj(\"./result/wnn_cellmask/inputs/default/predict/\", paste0(\"subset_\", subset_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "pp_dir <- \"./data/processed/wnn_cellmask/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Subset 1/8, File 1/24\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Subset 1/8, File 2/24\n",
      "\n",
      "Loading Subset 1/8, File 3/24\n",
      "\n",
      "Loading Subset 1/8, File 4/24\n",
      "\n",
      "Loading Subset 1/8, File 5/24\n",
      "\n",
      "Loading Subset 1/8, File 6/24\n",
      "\n",
      "Loading Subset 1/8, File 7/24\n",
      "\n",
      "Loading Subset 1/8, File 8/24\n",
      "\n",
      "Loading Subset 1/8, File 9/24\n",
      "\n",
      "Loading Subset 1/8, File 10/24\n",
      "\n",
      "Loading Subset 1/8, File 11/24\n",
      "\n",
      "Loading Subset 1/8, File 12/24\n",
      "\n",
      "Loading Subset 1/8, File 13/24\n",
      "\n",
      "Loading Subset 1/8, File 14/24\n",
      "\n",
      "Loading Subset 1/8, File 15/24\n",
      "\n",
      "Loading Subset 1/8, File 16/24\n",
      "\n",
      "Loading Subset 1/8, File 17/24\n",
      "\n",
      "Loading Subset 1/8, File 18/24\n",
      "\n",
      "Loading Subset 1/8, File 19/24\n",
      "\n",
      "Loading Subset 1/8, File 20/24\n",
      "\n",
      "Loading Subset 1/8, File 21/24\n",
      "\n",
      "Loading Subset 1/8, File 22/24\n",
      "\n",
      "Loading Subset 1/8, File 23/24\n",
      "\n",
      "Loading Subset 1/8, File 24/24\n",
      "\n",
      "Loading Subset 2/8, File 1/16\n",
      "\n",
      "Loading Subset 2/8, File 2/16\n",
      "\n",
      "Loading Subset 2/8, File 3/16\n",
      "\n",
      "Loading Subset 2/8, File 4/16\n",
      "\n",
      "Loading Subset 2/8, File 5/16\n",
      "\n",
      "Loading Subset 2/8, File 6/16\n",
      "\n",
      "Loading Subset 2/8, File 7/16\n",
      "\n",
      "Loading Subset 2/8, File 8/16\n",
      "\n",
      "Loading Subset 2/8, File 9/16\n",
      "\n",
      "Loading Subset 2/8, File 10/16\n",
      "\n",
      "Loading Subset 2/8, File 11/16\n",
      "\n",
      "Loading Subset 2/8, File 12/16\n",
      "\n",
      "Loading Subset 2/8, File 13/16\n",
      "\n",
      "Loading Subset 2/8, File 14/16\n",
      "\n",
      "Loading Subset 2/8, File 15/16\n",
      "\n",
      "Loading Subset 2/8, File 16/16\n",
      "\n",
      "Loading Subset 3/8, File 1/17\n",
      "\n",
      "Loading Subset 3/8, File 2/17\n",
      "\n",
      "Loading Subset 3/8, File 3/17\n",
      "\n",
      "Loading Subset 3/8, File 4/17\n",
      "\n",
      "Loading Subset 3/8, File 5/17\n",
      "\n",
      "Loading Subset 3/8, File 6/17\n",
      "\n",
      "Loading Subset 3/8, File 7/17\n",
      "\n",
      "Loading Subset 3/8, File 8/17\n",
      "\n",
      "Loading Subset 3/8, File 9/17\n",
      "\n",
      "Loading Subset 3/8, File 10/17\n",
      "\n",
      "Loading Subset 3/8, File 11/17\n",
      "\n",
      "Loading Subset 3/8, File 12/17\n",
      "\n",
      "Loading Subset 3/8, File 13/17\n",
      "\n",
      "Loading Subset 3/8, File 14/17\n",
      "\n",
      "Loading Subset 3/8, File 15/17\n",
      "\n",
      "Loading Subset 3/8, File 16/17\n",
      "\n",
      "Loading Subset 3/8, File 17/17\n",
      "\n",
      "Loading Subset 4/8, File 1/21\n",
      "\n",
      "Loading Subset 4/8, File 2/21\n",
      "\n",
      "Loading Subset 4/8, File 3/21\n",
      "\n",
      "Loading Subset 4/8, File 4/21\n",
      "\n",
      "Loading Subset 4/8, File 5/21\n",
      "\n",
      "Loading Subset 4/8, File 6/21\n",
      "\n",
      "Loading Subset 4/8, File 7/21\n",
      "\n",
      "Loading Subset 4/8, File 8/21\n",
      "\n",
      "Loading Subset 4/8, File 9/21\n",
      "\n",
      "Loading Subset 4/8, File 10/21\n",
      "\n",
      "Loading Subset 4/8, File 11/21\n",
      "\n",
      "Loading Subset 4/8, File 12/21\n",
      "\n",
      "Loading Subset 4/8, File 13/21\n",
      "\n",
      "Loading Subset 4/8, File 14/21\n",
      "\n",
      "Loading Subset 4/8, File 15/21\n",
      "\n",
      "Loading Subset 4/8, File 16/21\n",
      "\n",
      "Loading Subset 4/8, File 17/21\n",
      "\n",
      "Loading Subset 4/8, File 18/21\n",
      "\n",
      "Loading Subset 4/8, File 19/21\n",
      "\n",
      "Loading Subset 4/8, File 20/21\n",
      "\n",
      "Loading Subset 4/8, File 21/21\n",
      "\n",
      "Loading Subset 5/8, File 1/17\n",
      "\n",
      "Loading Subset 5/8, File 2/17\n",
      "\n",
      "Loading Subset 5/8, File 3/17\n",
      "\n",
      "Loading Subset 5/8, File 4/17\n",
      "\n",
      "Loading Subset 5/8, File 5/17\n",
      "\n",
      "Loading Subset 5/8, File 6/17\n",
      "\n",
      "Loading Subset 5/8, File 7/17\n",
      "\n",
      "Loading Subset 5/8, File 8/17\n",
      "\n",
      "Loading Subset 5/8, File 9/17\n",
      "\n",
      "Loading Subset 5/8, File 10/17\n",
      "\n",
      "Loading Subset 5/8, File 11/17\n",
      "\n",
      "Loading Subset 5/8, File 12/17\n",
      "\n",
      "Loading Subset 5/8, File 13/17\n",
      "\n",
      "Loading Subset 5/8, File 14/17\n",
      "\n",
      "Loading Subset 5/8, File 15/17\n",
      "\n",
      "Loading Subset 5/8, File 16/17\n",
      "\n",
      "Loading Subset 5/8, File 17/17\n",
      "\n",
      "Loading Subset 6/8, File 1/21\n",
      "\n",
      "Loading Subset 6/8, File 2/21\n",
      "\n",
      "Loading Subset 6/8, File 3/21\n",
      "\n",
      "Loading Subset 6/8, File 4/21\n",
      "\n",
      "Loading Subset 6/8, File 5/21\n",
      "\n",
      "Loading Subset 6/8, File 6/21\n",
      "\n",
      "Loading Subset 6/8, File 7/21\n",
      "\n",
      "Loading Subset 6/8, File 8/21\n",
      "\n",
      "Loading Subset 6/8, File 9/21\n",
      "\n",
      "Loading Subset 6/8, File 10/21\n",
      "\n",
      "Loading Subset 6/8, File 11/21\n",
      "\n",
      "Loading Subset 6/8, File 12/21\n",
      "\n",
      "Loading Subset 6/8, File 13/21\n",
      "\n",
      "Loading Subset 6/8, File 14/21\n",
      "\n",
      "Loading Subset 6/8, File 15/21\n",
      "\n",
      "Loading Subset 6/8, File 16/21\n",
      "\n",
      "Loading Subset 6/8, File 17/21\n",
      "\n",
      "Loading Subset 6/8, File 18/21\n",
      "\n",
      "Loading Subset 6/8, File 19/21\n",
      "\n",
      "Loading Subset 6/8, File 20/21\n",
      "\n",
      "Loading Subset 6/8, File 21/21\n",
      "\n",
      "Loading Subset 7/8, File 1/35\n",
      "\n",
      "Loading Subset 7/8, File 2/35\n",
      "\n",
      "Loading Subset 7/8, File 3/35\n",
      "\n",
      "Loading Subset 7/8, File 4/35\n",
      "\n",
      "Loading Subset 7/8, File 5/35\n",
      "\n",
      "Loading Subset 7/8, File 6/35\n",
      "\n",
      "Loading Subset 7/8, File 7/35\n",
      "\n",
      "Loading Subset 7/8, File 8/35\n",
      "\n",
      "Loading Subset 7/8, File 9/35\n",
      "\n",
      "Loading Subset 7/8, File 10/35\n",
      "\n",
      "Loading Subset 7/8, File 11/35\n",
      "\n",
      "Loading Subset 7/8, File 12/35\n",
      "\n",
      "Loading Subset 7/8, File 13/35\n",
      "\n",
      "Loading Subset 7/8, File 14/35\n",
      "\n",
      "Loading Subset 7/8, File 15/35\n",
      "\n",
      "Loading Subset 7/8, File 16/35\n",
      "\n",
      "Loading Subset 7/8, File 17/35\n",
      "\n",
      "Loading Subset 7/8, File 18/35\n",
      "\n",
      "Loading Subset 7/8, File 19/35\n",
      "\n",
      "Loading Subset 7/8, File 20/35\n",
      "\n",
      "Loading Subset 7/8, File 21/35\n",
      "\n",
      "Loading Subset 7/8, File 22/35\n",
      "\n",
      "Loading Subset 7/8, File 23/35\n",
      "\n",
      "Loading Subset 7/8, File 24/35\n",
      "\n",
      "Loading Subset 7/8, File 25/35\n",
      "\n",
      "Loading Subset 7/8, File 26/35\n",
      "\n",
      "Loading Subset 7/8, File 27/35\n",
      "\n",
      "Loading Subset 7/8, File 28/35\n",
      "\n",
      "Loading Subset 7/8, File 29/35\n",
      "\n",
      "Loading Subset 7/8, File 30/35\n",
      "\n",
      "Loading Subset 7/8, File 31/35\n",
      "\n",
      "Loading Subset 7/8, File 32/35\n",
      "\n",
      "Loading Subset 7/8, File 33/35\n",
      "\n",
      "Loading Subset 7/8, File 34/35\n",
      "\n",
      "Loading Subset 7/8, File 35/35\n",
      "\n",
      "Loading Subset 8/8, File 1/34\n",
      "\n",
      "Loading Subset 8/8, File 2/34\n",
      "\n",
      "Loading Subset 8/8, File 3/34\n",
      "\n",
      "Loading Subset 8/8, File 4/34\n",
      "\n",
      "Loading Subset 8/8, File 5/34\n",
      "\n",
      "Loading Subset 8/8, File 6/34\n",
      "\n",
      "Loading Subset 8/8, File 7/34\n",
      "\n",
      "Loading Subset 8/8, File 8/34\n",
      "\n",
      "Loading Subset 8/8, File 9/34\n",
      "\n",
      "Loading Subset 8/8, File 10/34\n",
      "\n",
      "Loading Subset 8/8, File 11/34\n",
      "\n",
      "Loading Subset 8/8, File 12/34\n",
      "\n",
      "Loading Subset 8/8, File 13/34\n",
      "\n",
      "Loading Subset 8/8, File 14/34\n",
      "\n",
      "Loading Subset 8/8, File 15/34\n",
      "\n",
      "Loading Subset 8/8, File 16/34\n",
      "\n",
      "Loading Subset 8/8, File 17/34\n",
      "\n",
      "Loading Subset 8/8, File 18/34\n",
      "\n",
      "Loading Subset 8/8, File 19/34\n",
      "\n",
      "Loading Subset 8/8, File 20/34\n",
      "\n",
      "Loading Subset 8/8, File 21/34\n",
      "\n",
      "Loading Subset 8/8, File 22/34\n",
      "\n",
      "Loading Subset 8/8, File 23/34\n",
      "\n",
      "Loading Subset 8/8, File 24/34\n",
      "\n",
      "Loading Subset 8/8, File 25/34\n",
      "\n",
      "Loading Subset 8/8, File 26/34\n",
      "\n",
      "Loading Subset 8/8, File 27/34\n",
      "\n",
      "Loading Subset 8/8, File 28/34\n",
      "\n",
      "Loading Subset 8/8, File 29/34\n",
      "\n",
      "Loading Subset 8/8, File 30/34\n",
      "\n",
      "Loading Subset 8/8, File 31/34\n",
      "\n",
      "Loading Subset 8/8, File 32/34\n",
      "\n",
      "Loading Subset 8/8, File 33/34\n",
      "\n",
      "Loading Subset 8/8, File 34/34\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "187162650"
      ],
      "text/latex": [
       "187162650"
      ],
      "text/markdown": [
       "187162650"
      ],
      "text/plain": [
       "[1] 187162650"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 46213\n"
     ]
    }
   ],
   "source": [
    "z_list <- list()\n",
    "cell_name_list <- list()\n",
    "label_list <- list()\n",
    "is_label <- T\n",
    "subset_name_list <- list()\n",
    "S <- length(subset_names)\n",
    "for (i in seq_along(subset_names)) {\n",
    "    subset_name <- subset_names[i]\n",
    "    z_dir    <- pj(input_dirs[i], \"x\", \"rna\")\n",
    "    fnames <- dir(path = z_dir, pattern = \".csv$\")\n",
    "    fnames <- str_sort(fnames, decreasing = F)\n",
    "    z_subset_list <- list()\n",
    "    N <- length(fnames)\n",
    "    for (n in seq_along(fnames)) {\n",
    "        message(paste0(\"Loading Subset \", i, \"/\", S, \", File \", n, \"/\", N))\n",
    "        z_subset_list[[n]] <- read.csv(file.path(z_dir, fnames[n]), header = F)\n",
    "    }\n",
    "    z_list[[subset_name]] <- bind_rows(z_subset_list)\n",
    "\n",
    "    cell_name_list[[subset_name]] <- read.csv(pj(pp_dir, paste0(\"subset_\", subset_ids[i]),\n",
    "        \"cell_names.csv\"), header = T)[, 2]\n",
    "    subset_name_list[[subset_name]] <- rep(subset_name, length(cell_name_list[[subset_name]]))\n",
    "}\n",
    "\n",
    "rna <- t(data.matrix(bind_rows(z_list))) # pseudo rna counts\n",
    "length(rna)\n",
    "print(length(do.call(\"c\", unname(cell_name_list))))\n",
    "colnames(rna) <- do.call(\"c\", unname(cell_name_list))\n",
    "rownames(rna) <- paste0(\"rna-\", seq_len(nrow(rna)))\n",
    "obj <- CreateSeuratObject(counts = rna, assay = \"rna\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Subset 1/8, File 1/24\n",
      "\n",
      "Loading Subset 1/8, File 2/24\n",
      "\n",
      "Loading Subset 1/8, File 3/24\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Subset 1/8, File 4/24\n",
      "\n",
      "Loading Subset 1/8, File 5/24\n",
      "\n",
      "Loading Subset 1/8, File 6/24\n",
      "\n",
      "Loading Subset 1/8, File 7/24\n",
      "\n",
      "Loading Subset 1/8, File 8/24\n",
      "\n",
      "Loading Subset 1/8, File 9/24\n",
      "\n",
      "Loading Subset 1/8, File 10/24\n",
      "\n",
      "Loading Subset 1/8, File 11/24\n",
      "\n",
      "Loading Subset 1/8, File 12/24\n",
      "\n",
      "Loading Subset 1/8, File 13/24\n",
      "\n",
      "Loading Subset 1/8, File 14/24\n",
      "\n",
      "Loading Subset 1/8, File 15/24\n",
      "\n",
      "Loading Subset 1/8, File 16/24\n",
      "\n",
      "Loading Subset 1/8, File 17/24\n",
      "\n",
      "Loading Subset 1/8, File 18/24\n",
      "\n",
      "Loading Subset 1/8, File 19/24\n",
      "\n",
      "Loading Subset 1/8, File 20/24\n",
      "\n",
      "Loading Subset 1/8, File 21/24\n",
      "\n",
      "Loading Subset 1/8, File 22/24\n",
      "\n",
      "Loading Subset 1/8, File 23/24\n",
      "\n",
      "Loading Subset 1/8, File 24/24\n",
      "\n",
      "Loading Subset 2/8, File 1/16\n",
      "\n",
      "Loading Subset 2/8, File 2/16\n",
      "\n",
      "Loading Subset 2/8, File 3/16\n",
      "\n",
      "Loading Subset 2/8, File 4/16\n",
      "\n",
      "Loading Subset 2/8, File 5/16\n",
      "\n",
      "Loading Subset 2/8, File 6/16\n",
      "\n",
      "Loading Subset 2/8, File 7/16\n",
      "\n",
      "Loading Subset 2/8, File 8/16\n",
      "\n",
      "Loading Subset 2/8, File 9/16\n",
      "\n",
      "Loading Subset 2/8, File 10/16\n",
      "\n",
      "Loading Subset 2/8, File 11/16\n",
      "\n",
      "Loading Subset 2/8, File 12/16\n",
      "\n",
      "Loading Subset 2/8, File 13/16\n",
      "\n",
      "Loading Subset 2/8, File 14/16\n",
      "\n",
      "Loading Subset 2/8, File 15/16\n",
      "\n",
      "Loading Subset 2/8, File 16/16\n",
      "\n",
      "Loading Subset 3/8, File 1/17\n",
      "\n",
      "Loading Subset 3/8, File 2/17\n",
      "\n",
      "Loading Subset 3/8, File 3/17\n",
      "\n",
      "Loading Subset 3/8, File 4/17\n",
      "\n",
      "Loading Subset 3/8, File 5/17\n",
      "\n",
      "Loading Subset 3/8, File 6/17\n",
      "\n",
      "Loading Subset 3/8, File 7/17\n",
      "\n",
      "Loading Subset 3/8, File 8/17\n",
      "\n",
      "Loading Subset 3/8, File 9/17\n",
      "\n",
      "Loading Subset 3/8, File 10/17\n",
      "\n",
      "Loading Subset 3/8, File 11/17\n",
      "\n",
      "Loading Subset 3/8, File 12/17\n",
      "\n",
      "Loading Subset 3/8, File 13/17\n",
      "\n",
      "Loading Subset 3/8, File 14/17\n",
      "\n",
      "Loading Subset 3/8, File 15/17\n",
      "\n",
      "Loading Subset 3/8, File 16/17\n",
      "\n",
      "Loading Subset 3/8, File 17/17\n",
      "\n",
      "Loading Subset 4/8, File 1/21\n",
      "\n",
      "Loading Subset 4/8, File 2/21\n",
      "\n",
      "Loading Subset 4/8, File 3/21\n",
      "\n",
      "Loading Subset 4/8, File 4/21\n",
      "\n",
      "Loading Subset 4/8, File 5/21\n",
      "\n",
      "Loading Subset 4/8, File 6/21\n",
      "\n",
      "Loading Subset 4/8, File 7/21\n",
      "\n",
      "Loading Subset 4/8, File 8/21\n",
      "\n",
      "Loading Subset 4/8, File 9/21\n",
      "\n",
      "Loading Subset 4/8, File 10/21\n",
      "\n",
      "Loading Subset 4/8, File 11/21\n",
      "\n",
      "Loading Subset 4/8, File 12/21\n",
      "\n",
      "Loading Subset 4/8, File 13/21\n",
      "\n",
      "Loading Subset 4/8, File 14/21\n",
      "\n",
      "Loading Subset 4/8, File 15/21\n",
      "\n",
      "Loading Subset 4/8, File 16/21\n",
      "\n",
      "Loading Subset 4/8, File 17/21\n",
      "\n",
      "Loading Subset 4/8, File 18/21\n",
      "\n",
      "Loading Subset 4/8, File 19/21\n",
      "\n",
      "Loading Subset 4/8, File 20/21\n",
      "\n",
      "Loading Subset 4/8, File 21/21\n",
      "\n",
      "Loading Subset 5/8, File 1/17\n",
      "\n",
      "Loading Subset 5/8, File 2/17\n",
      "\n",
      "Loading Subset 5/8, File 3/17\n",
      "\n",
      "Loading Subset 5/8, File 4/17\n",
      "\n",
      "Loading Subset 5/8, File 5/17\n",
      "\n",
      "Loading Subset 5/8, File 6/17\n",
      "\n",
      "Loading Subset 5/8, File 7/17\n",
      "\n",
      "Loading Subset 5/8, File 8/17\n",
      "\n",
      "Loading Subset 5/8, File 9/17\n",
      "\n",
      "Loading Subset 5/8, File 10/17\n",
      "\n",
      "Loading Subset 5/8, File 11/17\n",
      "\n",
      "Loading Subset 5/8, File 12/17\n",
      "\n",
      "Loading Subset 5/8, File 13/17\n",
      "\n",
      "Loading Subset 5/8, File 14/17\n",
      "\n",
      "Loading Subset 5/8, File 15/17\n",
      "\n",
      "Loading Subset 5/8, File 16/17\n",
      "\n",
      "Loading Subset 5/8, File 17/17\n",
      "\n",
      "Loading Subset 6/8, File 1/21\n",
      "\n",
      "Loading Subset 6/8, File 2/21\n",
      "\n",
      "Loading Subset 6/8, File 3/21\n",
      "\n",
      "Loading Subset 6/8, File 4/21\n",
      "\n",
      "Loading Subset 6/8, File 5/21\n",
      "\n",
      "Loading Subset 6/8, File 6/21\n",
      "\n",
      "Loading Subset 6/8, File 7/21\n",
      "\n",
      "Loading Subset 6/8, File 8/21\n",
      "\n",
      "Loading Subset 6/8, File 9/21\n",
      "\n",
      "Loading Subset 6/8, File 10/21\n",
      "\n",
      "Loading Subset 6/8, File 11/21\n",
      "\n",
      "Loading Subset 6/8, File 12/21\n",
      "\n",
      "Loading Subset 6/8, File 13/21\n",
      "\n",
      "Loading Subset 6/8, File 14/21\n",
      "\n",
      "Loading Subset 6/8, File 15/21\n",
      "\n",
      "Loading Subset 6/8, File 16/21\n",
      "\n",
      "Loading Subset 6/8, File 17/21\n",
      "\n",
      "Loading Subset 6/8, File 18/21\n",
      "\n",
      "Loading Subset 6/8, File 19/21\n",
      "\n",
      "Loading Subset 6/8, File 20/21\n",
      "\n",
      "Loading Subset 6/8, File 21/21\n",
      "\n",
      "Loading Subset 7/8, File 1/35\n",
      "\n",
      "Loading Subset 7/8, File 2/35\n",
      "\n",
      "Loading Subset 7/8, File 3/35\n",
      "\n",
      "Loading Subset 7/8, File 4/35\n",
      "\n",
      "Loading Subset 7/8, File 5/35\n",
      "\n",
      "Loading Subset 7/8, File 6/35\n",
      "\n",
      "Loading Subset 7/8, File 7/35\n",
      "\n",
      "Loading Subset 7/8, File 8/35\n",
      "\n",
      "Loading Subset 7/8, File 9/35\n",
      "\n",
      "Loading Subset 7/8, File 10/35\n",
      "\n",
      "Loading Subset 7/8, File 11/35\n",
      "\n",
      "Loading Subset 7/8, File 12/35\n",
      "\n",
      "Loading Subset 7/8, File 13/35\n",
      "\n",
      "Loading Subset 7/8, File 14/35\n",
      "\n",
      "Loading Subset 7/8, File 15/35\n",
      "\n",
      "Loading Subset 7/8, File 16/35\n",
      "\n",
      "Loading Subset 7/8, File 17/35\n",
      "\n",
      "Loading Subset 7/8, File 18/35\n",
      "\n",
      "Loading Subset 7/8, File 19/35\n",
      "\n",
      "Loading Subset 7/8, File 20/35\n",
      "\n",
      "Loading Subset 7/8, File 21/35\n",
      "\n",
      "Loading Subset 7/8, File 22/35\n",
      "\n",
      "Loading Subset 7/8, File 23/35\n",
      "\n",
      "Loading Subset 7/8, File 24/35\n",
      "\n",
      "Loading Subset 7/8, File 25/35\n",
      "\n",
      "Loading Subset 7/8, File 26/35\n",
      "\n",
      "Loading Subset 7/8, File 27/35\n",
      "\n",
      "Loading Subset 7/8, File 28/35\n",
      "\n",
      "Loading Subset 7/8, File 29/35\n",
      "\n",
      "Loading Subset 7/8, File 30/35\n",
      "\n",
      "Loading Subset 7/8, File 31/35\n",
      "\n",
      "Loading Subset 7/8, File 32/35\n",
      "\n",
      "Loading Subset 7/8, File 33/35\n",
      "\n",
      "Loading Subset 7/8, File 34/35\n",
      "\n",
      "Loading Subset 7/8, File 35/35\n",
      "\n",
      "Loading Subset 8/8, File 1/34\n",
      "\n",
      "Loading Subset 8/8, File 2/34\n",
      "\n",
      "Loading Subset 8/8, File 3/34\n",
      "\n",
      "Loading Subset 8/8, File 4/34\n",
      "\n",
      "Loading Subset 8/8, File 5/34\n",
      "\n",
      "Loading Subset 8/8, File 6/34\n",
      "\n",
      "Loading Subset 8/8, File 7/34\n",
      "\n",
      "Loading Subset 8/8, File 8/34\n",
      "\n",
      "Loading Subset 8/8, File 9/34\n",
      "\n",
      "Loading Subset 8/8, File 10/34\n",
      "\n",
      "Loading Subset 8/8, File 11/34\n",
      "\n",
      "Loading Subset 8/8, File 12/34\n",
      "\n",
      "Loading Subset 8/8, File 13/34\n",
      "\n",
      "Loading Subset 8/8, File 14/34\n",
      "\n",
      "Loading Subset 8/8, File 15/34\n",
      "\n",
      "Loading Subset 8/8, File 16/34\n",
      "\n",
      "Loading Subset 8/8, File 17/34\n",
      "\n",
      "Loading Subset 8/8, File 18/34\n",
      "\n",
      "Loading Subset 8/8, File 19/34\n",
      "\n",
      "Loading Subset 8/8, File 20/34\n",
      "\n",
      "Loading Subset 8/8, File 21/34\n",
      "\n",
      "Loading Subset 8/8, File 22/34\n",
      "\n",
      "Loading Subset 8/8, File 23/34\n",
      "\n",
      "Loading Subset 8/8, File 24/34\n",
      "\n",
      "Loading Subset 8/8, File 25/34\n",
      "\n",
      "Loading Subset 8/8, File 26/34\n",
      "\n",
      "Loading Subset 8/8, File 27/34\n",
      "\n",
      "Loading Subset 8/8, File 28/34\n",
      "\n",
      "Loading Subset 8/8, File 29/34\n",
      "\n",
      "Loading Subset 8/8, File 30/34\n",
      "\n",
      "Loading Subset 8/8, File 31/34\n",
      "\n",
      "Loading Subset 8/8, File 32/34\n",
      "\n",
      "Loading Subset 8/8, File 33/34\n",
      "\n",
      "Loading Subset 8/8, File 34/34\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "10351712"
      ],
      "text/latex": [
       "10351712"
      ],
      "text/markdown": [
       "10351712"
      ],
      "text/plain": [
       "[1] 10351712"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 46213\n"
     ]
    }
   ],
   "source": [
    "z_list <- list()\n",
    "cell_name_list <- list()\n",
    "label_list <- list()\n",
    "is_label <- T\n",
    "subset_name_list <- list()\n",
    "S <- length(subset_names)\n",
    "for (i in seq_along(subset_names)) {\n",
    "    subset_name <- subset_names[i]\n",
    "    z_dir    <- pj(input_dirs[i], \"x\", \"adt\")\n",
    "    fnames <- dir(path = z_dir, pattern = \".csv$\")\n",
    "    fnames <- str_sort(fnames, decreasing = F)\n",
    "    z_subset_list <- list()\n",
    "    N <- length(fnames)\n",
    "    for (n in seq_along(fnames)) {\n",
    "        message(paste0(\"Loading Subset \", i, \"/\", S, \", File \", n, \"/\", N))\n",
    "        z_subset_list[[n]] <- read.csv(file.path(z_dir, fnames[n]), header = F)\n",
    "    }\n",
    "    z_list[[subset_name]] <- bind_rows(z_subset_list)\n",
    "\n",
    "    cell_name_list[[subset_name]] <- read.csv(pj(pp_dir, paste0(\"subset_\", subset_ids[i]),\n",
    "        \"cell_names.csv\"), header = T)[, 2]\n",
    "    subset_name_list[[subset_name]] <- rep(subset_name, length(cell_name_list[[subset_name]]))\n",
    "}\n",
    "\n",
    "rna <- t(data.matrix(bind_rows(z_list)))  # pseudo rna counts\n",
    "length(rna)\n",
    "print(length(do.call(\"c\", unname(cell_name_list))))\n",
    "colnames(rna) <- do.call(\"c\", unname(cell_name_list))\n",
    "rownames(rna) <- paste0(\"rna-\", seq_len(nrow(rna)))\n",
    "obj_adt <- CreateSeuratObject(counts = rna, assay = \"adt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Non-unique features (rownames) present in the input matrix, making unique\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "4274 features across 46213 samples within 1 assay \n",
       "Active assay: RNA (4274 features, 0 variable features)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "adata <- cat_data(obj, obj_adt)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "4274 features across 6142 samples within 1 assay \n",
       "Active assay: RNA (4274 features, 0 variable features)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref_adata <- adata[, adata[['orig.ident']]=='B0']\n",
    "ref_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing\n",
      "\n",
      "Finding variable genes using vst method\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Total 3617 genes for downstream steps\n",
      "\n",
      "Scaling and PCA\n",
      "\n",
      "Running Harmony integration\n",
      "\n",
      "Harmony 1/20\n",
      "\n",
      "Harmony 2/20\n",
      "\n",
      "Harmony converged after 2 iterations\n",
      "\n",
      "Computing reference compression terms\n",
      "\n",
      "Running UMAP\n",
      "\n",
      "File already exists at that path... overwriting...\n",
      "\n",
      "Saved uwot model\n",
      "\n"
     ]
    }
   ],
   "source": [
    "reference = symphony::buildReference(\n",
    "    ref_adata@assays$RNA@counts,                   # reference expression (genes by cells)\n",
    "    ref_adata@meta.data,              # reference metadata (cells x attributes)\n",
    "    vars = c('orig.ident'),         # variable(s) to integrate over\n",
    "    K = 100,                   # number of Harmony soft clusters\n",
    "    verbose = TRUE,            # display verbose output\n",
    "    do_umap = TRUE,            # run UMAP and save UMAP model to file\n",
    "    do_normalize = TRUE,      # perform log(CP10k) normalization on reference expression\n",
    "    vargenes_method = 'vst',   # variable gene selection method: 'vst' or 'mvp'\n",
    "    vargenes_groups = 'orig.ident', # metadata column specifying groups for variable gene selection within each group\n",
    "    topn = 3617,               # number of variable genes (per group)\n",
    "    theta = 2,                 # Harmony parameter(s) for diversity term\n",
    "    d = 32,                    # number of dimensions for PCA\n",
    "    save_uwot_path = './result/wnn_cellmask/symphony_obo/default/train/', # file path to save uwot UMAP model\n",
    "    additional_genes = NULL    # vector of any additional genes to force include\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "write.table(t(reference$Z_corr), './result/wnn_cellmask/symphony_obo/default/predict/subset_0/z/joint/00.csv', sep=',', row.names =F, col.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# umap_labels = cbind(ref_adata@meta.data, reference$umap$embedding)\n",
    "# fig.size(3, 5)\n",
    "# plotBasic(umap_labels, title = 'Reference', color.by = 'l1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing\n",
      "\n",
      "Scaling and synchronizing query gene expression\n",
      "\n",
      "Found 3617 out of 3617 reference variable genes in query dataset\n",
      "\n",
      "Project query cells using reference gene loadings\n",
      "\n",
      "Clustering query cells to reference centroids\n",
      "\n",
      "Correcting query batch effects\n",
      "\n",
      "All done!\n",
      "\n",
      "Normalizing\n",
      "\n",
      "Scaling and synchronizing query gene expression\n",
      "\n",
      "Found 3617 out of 3617 reference variable genes in query dataset\n",
      "\n",
      "Project query cells using reference gene loadings\n",
      "\n",
      "Clustering query cells to reference centroids\n",
      "\n",
      "Correcting query batch effects\n",
      "\n",
      "All done!\n",
      "\n",
      "Normalizing\n",
      "\n",
      "Scaling and synchronizing query gene expression\n",
      "\n",
      "Found 3617 out of 3617 reference variable genes in query dataset\n",
      "\n",
      "Project query cells using reference gene loadings\n",
      "\n",
      "Clustering query cells to reference centroids\n",
      "\n",
      "Correcting query batch effects\n",
      "\n",
      "All done!\n",
      "\n",
      "Normalizing\n",
      "\n",
      "Scaling and synchronizing query gene expression\n",
      "\n",
      "Found 3617 out of 3617 reference variable genes in query dataset\n",
      "\n",
      "Project query cells using reference gene loadings\n",
      "\n",
      "Clustering query cells to reference centroids\n",
      "\n",
      "Correcting query batch effects\n",
      "\n",
      "All done!\n",
      "\n",
      "Normalizing\n",
      "\n",
      "Scaling and synchronizing query gene expression\n",
      "\n",
      "Found 3617 out of 3617 reference variable genes in query dataset\n",
      "\n",
      "Project query cells using reference gene loadings\n",
      "\n",
      "Clustering query cells to reference centroids\n",
      "\n",
      "Correcting query batch effects\n",
      "\n",
      "All done!\n",
      "\n",
      "Normalizing\n",
      "\n",
      "Scaling and synchronizing query gene expression\n",
      "\n",
      "Found 3617 out of 3617 reference variable genes in query dataset\n",
      "\n",
      "Project query cells using reference gene loadings\n",
      "\n",
      "Clustering query cells to reference centroids\n",
      "\n",
      "Correcting query batch effects\n",
      "\n",
      "All done!\n",
      "\n",
      "Normalizing\n",
      "\n",
      "Scaling and synchronizing query gene expression\n",
      "\n",
      "Found 3617 out of 3617 reference variable genes in query dataset\n",
      "\n",
      "Project query cells using reference gene loadings\n",
      "\n",
      "Clustering query cells to reference centroids\n",
      "\n",
      "Correcting query batch effects\n",
      "\n",
      "All done!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(i in c(1:7)){\n",
    "    query_adata <- adata[, adata[['orig.ident']]==paste0('B', i)]\n",
    "    query = mapQuery(query_adata@assays$RNA@counts,             # query gene expression (genes x cells)\n",
    "                 query_adata@meta.data,        # query metadata (cells x attributes)\n",
    "                 reference,             # Symphony reference object\n",
    "                 vars = NULL,           # Query batch variables to harmonize over (NULL treats query as one batch)\n",
    "                 do_normalize = TRUE,  # perform log(CP10k) normalization on query (set to FALSE if already normalized)\n",
    "                 do_umap = FALSE)\n",
    "    dir.create(paste0('./MIRACLE/result/wnn_cellmask/symphony_obo/default/predict/subset_',i,'/z/joint/'), recursive=T)\n",
    "    write.table(t(query$Z), paste0('./MIRACLE/result/wnn_cellmask/symphony_obo/default/predict/subset_',i,'/z/joint/00.csv'), sep=',', row.names =F, col.names = F)\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# offline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing ADT with data\n",
      "\n",
      "Adding counts for ADT\n",
      "\n",
      "Adding miscellaneous information for ADT\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "3841 features across 46326 samples within 1 assay \n",
       "Active assay: RNA (3841 features, 0 variable features)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref_adata <- adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dir.create('./results/wnn_cellmask/symphony_offline/default/train', recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing\n",
      "\n",
      "Finding variable genes using vst method\n",
      "\n",
      "Total 3841 genes for downstream steps\n",
      "\n",
      "Scaling and PCA\n",
      "\n",
      "Running Harmony integration\n",
      "\n",
      "Harmony 1/20\n",
      "\n",
      "Harmony 2/20\n",
      "\n",
      "Harmony 3/20\n",
      "\n",
      "Harmony 4/20\n",
      "\n",
      "Harmony 5/20\n",
      "\n",
      "Harmony converged after 5 iterations\n",
      "\n",
      "Computing reference compression terms\n",
      "\n",
      "Running UMAP\n",
      "\n",
      "Saved uwot model\n",
      "\n"
     ]
    }
   ],
   "source": [
    "reference = symphony::buildReference(\n",
    "    adata@assays$RNA@counts,                   # reference expression (genes by cells)\n",
    "    adata@meta.data,              # reference metadata (cells x attributes)\n",
    "    vars = c('orig.ident'),         # variable(s) to integrate over\n",
    "    K = 100,                   # number of Harmony soft clusters\n",
    "    verbose = TRUE,            # display verbose output\n",
    "    do_umap = TRUE,            # run UMAP and save UMAP model to file\n",
    "    do_normalize = TRUE,      # perform log(CP10k) normalization on reference expression\n",
    "    vargenes_method = 'vst',   # variable gene selection method: 'vst' or 'mvp'\n",
    "    vargenes_groups = 'orig.ident', # metadata column specifying groups for variable gene selection within each group\n",
    "    topn = 3841,               # number of variable genes (per group)\n",
    "    theta = 2,                 # Harmony parameter(s) for diversity term\n",
    "    d = 32,                    # number of dimensions for PCA\n",
    "    save_uwot_path = './results/wnn_cellmask/symphony_offline/default/train/uwot_model_1', # file path to save uwot UMAP model\n",
    "    additional_genes = NULL    # vector of any additional genes to force include\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "for(i in c(0:7)){\n",
    "    a <- reference[, reference[['orig.ident']]==paste0('B', i)]\n",
    "    dir.create(paste0('./results/wnn_cellmask/symphony_offline/default/predict/subset_',i), recursive = T)\n",
    "    write.table(t(a$Z_corr), paste0('./results/wnn_cellmask/symphony_offline/default/predict/subset_',i,'00.csv'), sep=',', row.names =F, col.names = F)\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
